File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.py
def solution(n, t, m, timetable):
    crew = []
    for time in timetable:
        hour, min = map(int, time.split(':'))
        crew.append(hour*60+min)
    crew.sort()
    crew_idx = 0
    last_ridden = 0
    last_time = 0
    
    for i in range(n):
        last_bus_time= 9 * 60 + (t * i)
        ridden = 0

        while crew_idx < len(crew) and crew[crew_idx] <= last_bus_time and ridden < m:
            last_time = crew[crew_idx]
            crew_idx += 1
            ridden += 1
        
    if ridden < m:
        result_time = last_bus_time 
    else:
        result_time = max(0, last_time - 1)

    H, M = divmod(result_time, 60)
    return f'{H:02d}:{M:02d}'

File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.cpp
#include<bits/stdc++.h>
#include <vector>
#include <string>
using namespace std;

string solution(int n, int t, int m, vector<string> timetable) {
    vector<int> crew;
    for (const string& time : timetable){
        int hour = stoi(time.substr(0, 2));
        int min = stoi(time.substr(3, 2));
        crew.push_back(hour * 60 + min);
    }

    sort(crew.begin(),crew.end());
    int crew_idx = 0;
    int last_crew_time = 0;
    int ridden;
    int last_bus_time;

    for (int i = 0; i < n; i++){
        last_bus_time = 9 * 60 + t * i;
        ridden = 0;

        while (crew_idx < crew.size() && crew[crew_idx] <= last_bus_time && ridden < m){
            last_crew_time = crew[crew_idx];
            crew_idx++;
            ridden++;
        }
    }

    int answer = 0;
    if (ridden < m) answer = last_bus_time;
    else answer = last_crew_time - 1;
    
    int H = answer / 60;
    int M = answer % 60;

    ostringstream oss;
    oss << setw(2) << setfill('0') << H
        << ":" 
        << setw(2) << setfill('0') << M;

    return oss.str(); 
}